### PR TITLE
datanode selector filter, sort and search

### LINKED ISSUE #824: Data Node Selector - Custom filter
## Improvement

It can be extremely useful to display only some data nodes (some can be irrelevant) and have a custom user function to handle this filter. 

Example: The selector could use a function for each Data Node, which could return True to display the Data Node in the Data Node Selector.

Example of code:
```
<|{datanode}|data_node_selector|filter_scenario={scenario}|filter={my_filter}|>
```

Example of function:

```python
def my_filter(data_node:DataNode) -> bool:
    return data_node.config_id == "TOTO"
```

Note: The generic and scenario filters should be applied with an AND operator.

Taipy version:
Taipy 3.0

### LINKED ISSUE #1352: [Refactor] Data Node Selector - Default filters should be consistent with config
### Description

The goal of this issue is to improve the UX of Data Node Selector filters. Some filters are for example not useful when not using validity period:

- expired
- expiration date

These filters should not be shown as they were not used in this case.

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Propagate any change on the demos and run all of them to ensure there is no breaking change.
- [ ] Ensure any change is well documented.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### LINKED ISSUE #1373: [🐛 BUG] Scenario Selector - Filter on creation date not working
### What went wrong? 🤔

Filtering on the creation date does not work.

```
C:\Users\jacta\.conda\envs\develop\lib\site-packages\taipy\gui_core\_adapters.py:249: TaipyGuiWarning: Error filtering with creation_date >= 2024-06-19 22:00:00 on Thursday, 27. June 2024 > g.:
'>=' not supported between instances of 'str' and 'datetime.datetime'
  _warn(f"Error filtering with {col} {action} {val} on {ent}.", e)
```

No filter is applied.

### Expected Behavior

The filter should be applied correctly.

### Steps to Reproduce Issue

```python


# Import necessary libraries
import pandas as pd
import taipy as tp
from taipy import Config, Scope, Frequency

import datetime as dt


# Function to run a Dataiku scenario
def run_something(input_1, input_2):
    datetime = dt.datetime.now()
    date = dt.date(2018, 1, 1)
    int_var = 10
    string_var = "String"
    return datetime, date, int_var, string_var


data = {"toto": [i for i in range(10_000)], 
        "titi": [2*i for i in range(10_000)],
        "tata": [4*i for i in range(10_000)]}



input_1_cfg = Config.configure_data_node(
    id="input_1_data_node",
    default_data=data,
)


input_2_cfg = Config.configure_data_node(
    id="input_2_data_node",
    default_data=data,
)

datetime_cfg = Config.configure_data_node(id="datetime_data_node")
date_cfg = Config.configure_data_node(id="date_data_node")
int_cfg = Config.configure_data_node(id="int_data_node")
string_cfg = Config.configure_data_node(id="string_data_node")


# Scenario and task configuration in Taipy
scenario_task_cfg = Config.configure_task(
    id="scenario_task",
    function=run_something,
    input=[input_1_cfg, input_2_cfg],
    output=[datetime_cfg, date_cfg, int_cfg, string_cfg]
)

scenario_cfg = Config.configure_scenario(
    id="scenario",
    task_configs=[scenario_task_cfg],
    frequency=Frequency.DAILY)

data_node = None

# GUI Markdown content
scenario_md = """
<|1 1|layout|
<|{scenario}|scenario_selector|>

<|{scenario}|scenario|>
|>

<|{scenario}|>


<|job_selector|>

<|1 1|layout|
Input 1 Dataset:
<|{scenario.input_1_data_node if scenario else None}|data_node|>

Input 2 Dataset:
<|{scenario.input_2_data_node if scenario else None}|data_node|>
|>


<|1 1|layout|
<|{data_node}|data_node_selector|>

<|{data_node}|data_node|>
|>
"""

# Main execution block with GUI setup
if __name__ == "__main__":
    tp.Core().run()
    scenario = tp.create_scenario(scenario_cfg)
    # Run the GUI
    tp.Gui(scenario_md).run(title="Bug replication", port=3248)

```

### Version of Taipy

Develop - 6/7/2024 (M/D/Y)

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### LINKED ISSUE #1374: [🐛 BUG] Scenario Selector - Filter on Cycle not working
### What went wrong? 🤔

Filtering on the cycle start does not work.

```
C:\Users\jacta\.conda\envs\develop\lib\site-packages\taipy\gui_core\_adapters.py:249: TaipyGuiWarning: Error filtering with cycle.start >= 2024-06-19 22:00:00 on Thursday, 27. June 2024 > g.:
start is not an attribute of cycle CYCLE_Thursday_27._June_2024_dad68cc2-b8a5-4841-a4ca-6dfc107542b3        
  _warn(f"Error filtering with {col} {action} {val} on {ent}.", e)
```

The filter on the cycle end does not work.

```
C:\Users\jacta\.conda\envs\develop\lib\site-packages\taipy\gui_core\_adapters.py:249: TaipyGuiWarning: Error filtering with cycle.end >= 2024-06-19 22:00:00 on Thursday, 27. June 2024 > g.:
end is not an attribute of cycle CYCLE_Thursday_27._June_2024_dad68cc2-b8a5-4841-a4ca-6dfc107542b3
  _warn(f"Error filtering with {col} {action} {val} on {ent}.", e)
```
No filter is applied.

### Expected Behavior

The filters should be applied correctly.

### Steps to Reproduce Issue

```python
# Import necessary libraries
import pandas as pd
import taipy as tp
from taipy import Config, Scope, Frequency

import datetime as dt


# Function to run a Dataiku scenario
def run_something(input_1, input_2):
    datetime = dt.datetime.now()
    date = dt.date(2018, 1, 1)
    int_var = 10
    string_var = "String"
    return datetime, date, int_var, string_var


data = {"toto": [i for i in range(10_000)], 
        "titi": [2*i for i in range(10_000)],
        "tata": [4*i for i in range(10_000)]}



input_1_cfg = Config.configure_data_node(
    id="input_1_data_node",
    default_data=data,
)


input_2_cfg = Config.configure_data_node(
    id="input_2_data_node",
    default_data=data,
)

datetime_cfg = Config.configure_data_node(id="datetime_data_node")
date_cfg = Config.configure_data_node(id="date_data_node")
int_cfg = Config.configure_data_node(id="int_data_node")
string_cfg = Config.configure_data_node(id="string_data_node")


# Scenario and task configuration in Taipy
scenario_task_cfg = Config.configure_task(
    id="scenario_task",
    function=run_something,
    input=[input_1_cfg, input_2_cfg],
    output=[datetime_cfg, date_cfg, int_cfg, string_cfg]
)

scenario_cfg = Config.configure_scenario(
    id="scenario",
    task_configs=[scenario_task_cfg],
    frequency=Frequency.DAILY)

data_node = None

# GUI Markdown content
scenario_md = """
<|1 1|layout|
<|{scenario}|scenario_selector|>

<|{scenario}|scenario|>
|>

<|{scenario}|>


<|job_selector|>

<|1 1|layout|
Input 1 Dataset:
<|{scenario.input_1_data_node if scenario else None}|data_node|>

Input 2 Dataset:
<|{scenario.input_2_data_node if scenario else None}|data_node|>
|>


<|1 1|layout|
<|{data_node}|data_node_selector|>

<|{data_node}|data_node|>
|>
"""

# Main execution block with GUI setup
if __name__ == "__main__":
    tp.Core().run()
    scenario = tp.create_scenario(scenario_cfg)
    # Run the GUI
    tp.Gui(scenario_md).run(title="Bug replication", port=3248)

```

### Version of Taipy

Develop - 6/6/2024

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by FredLL-Avaiga**
resolves #1352 
resolves #824 
resolves #1373 
resolves #1374

### CODE REVIEW

**Review by dinhlongviolin1 — DISMISSED** [review]
LGTM

File: frontend/taipy/src/CoreSelector.tsx [hunk 1]
```diff
@@ -526,15 +527,16 @@ const CoreSelector = (props: CoreSelectorProps) => {
                 if (old.length != filters.length || JSON.stringify(old) != jsonFilters) {
                     localStoreSet(jsonFilters, id, lovPropertyName, "filter");
                     const filterVar = getUpdateVar(updateCoreVars, "filter");
-                    dispatch(
+                    const lovVar = getUpdateVarNames(updateVars, lovPropertyName);
```
Comment (FabienLelaquais) [inline frontend/taipy/src/CoreSelector.tsx:528]: I suspect a comment here would be more than welcome.
Why do you want to delay the dispatch()...

File: taipy/gui_core/_GuiCoreLib.py [hunk 2]
```diff
@@ -165,15 +171,29 @@ class _GuiCore(ElementLibrary):
                 "class_name": ElementProperty(PropertyType.dynamic_string),
                 "show_pins": ElementProperty(PropertyType.boolean, True),
                 __DATANODE_SEL_SCENARIO_PROP: ElementProperty(PropertyType.dynamic_list),
+                __SEL_DATANODES_PROP: ElementProperty(PropertyType.dynamic_list),
                 "multiple": ElementProperty(PropertyType.boolean, False),
+                "filter": ElementProperty(_GuiCoreDatanodeFilter, "*"),
+                "sort": ElementProperty(_GuiCoreDatanodeSort, "*"),
```
Comment (FabienLelaquais) [inline taipy/gui_core/_GuiCoreLib.py:177]: sort = "*"?
Wouldn't None be better?

File: taipy/gui_core/_adapters.py [hunk 3]
```diff
@@ -319,30 +342,82 @@ def get(self):
...
+                    (attr, self.get_type(attr), self.get_enums().get(attr)) if self.full_desc() else (attr,)
                     for attr in flist
                     if attr and isinstance(attr, str)
                 ]
             )
         return None
 
 
-class _GuiCoreScenarioFilter(_GuiCoreScenarioProperties):
+class _GuiCorePropDesc:
```
Comment (FabienLelaquais) [inline taipy/gui_core/_adapters.py:357]: Make this a dataclass?

File: taipy/gui_core/_adapters.py [hunk 4]
```diff
@@ -235,63 +246,72 @@ def get_hash():
...
     except Exception as e:
-        _warn(f"Error filtering with {col} {action} {val} on {ent}.", e)
+        if _is_debugging():
+            _warn(f"Error filtering with {col} {action} {val} on {ent}.", e)
+        return col_type == "any" and action == "!="
     return True
 
 
+def _get_entity_property(col: str, a_type: t.Type):
+    col_parts = col.split("(")
```
Comment (FabienLelaquais) [inline taipy/gui_core/_adapters.py]: Really surprised by this splitting on '('...
Please add a comment on the function col parameter.

**Review by FabienLelaquais — DISMISSED** [review]

File: taipy/gui_core/_GuiCoreLib.py (hunk 2 above)
Comment (FredLL-Avaiga) [reply taipy/gui_core/_GuiCoreLib.py:177]: it's coherent with the doc: a list by default

File: taipy/gui_core/_adapters.py (hunk 3 above)
Comment (FredLL-Avaiga) [reply taipy/gui_core/_adapters.py:357]: I thought of this but I forgot, on it

**Review by FabienLelaquais — APPROVED** [review]
👌

**Pull request merged**